### PR TITLE
ROX-9398: auto-delete image integrations associate with a deleted cluster

### PR DESCRIPTION
**Pull request by daynewlee**
## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why
you did not do so. Valid reasons include, for example, "CI is sufficient",
"No testable changes". Feel free to attach JSON snippets, curl commands,
screenshots.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: central/cluster/datastore/datastore_impl.go [hunk 1]
```diff
@@ -152,6 +154,23 @@ func (ds *datastoreImpl) UpdateClusterStatus(ctx context.Context, id string, sta
...
+
+	q := pkgSearch.NewQueryBuilder().AddExactMatches(pkgSearch.ClusterID, cluster.GetId()).ProtoQuery()
+
+	imageIntegrations, err := ds.imageIntegrationDataStore.Search(ctx, q)
+	if err != nil {
+		log.Errorf("failed to get image integrations for removed cluster %s: %v", cluster.GetId(), err)
+		return
+	}
+	for _, imageIntegration := range imageIntegrations {
+		err = ds.imageIntegrationDataStore.RemoveImageIntegration(ctx, imageIntegration.ID)
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply central/cluster/datastore/datastore_impl.go:167]: likewise in the comment below

File: central/imageintegration/datastore/datastore.go [hunk 2]
```diff
@@ -16,17 +25,50 @@ var (
 // DataStore is the entry point for modifying Cluster data.
 //go:generate mockgen-wrapper
 type DataStore interface {
-	GetImageIntegration(ctx context.Context, id string) (*storage.ImageIntegration, bool, error)
-	GetImageIntegrations(ctx context.Context, integration *v1.GetImageIntegrationsRequest) ([]*storage.ImageIntegration, error)
+	GetImageIntegration(ctx context.Context, id string) (*genSt.ImageIntegration, bool, error)
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore.go:28]: call this storage and change the var name that is conflicts with

File: central/imageintegration/datastore/datastore.go [hunk 3]
```diff
@@ -16,17 +25,50 @@ var (
...
-	return &datastoreImpl{
-		storage: storage,
+func New(storage store.Store, indexer index.Indexer, searcher search.Searcher) DataStore {
+	ds := &datastoreImpl{
+		storage:           storage,
+		indexer:           indexer,
+		formattedSearcher: searcher,
 	}
+
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(),
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore.go:45]: since internal, let's just use all access

File: central/imageintegration/datastore/datastore.go [hunk 4]
```diff
@@ -16,17 +25,50 @@ var (
...
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(genSt.Access_READ_ACCESS, genSt.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.ImageIntegration)))
+
+	if err := ds.buildIndex(ctx); err != nil {
+		log.Error(err.Error())
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore.go:51]: We need to have new return an err and fail on that. This will cause a confusing panic

File: central/imageintegration/datastore/datastore_impl.go [hunk 5]
```diff
@@ -65,7 +74,11 @@ func (ds *datastoreImpl) AddImageIntegration(ctx context.Context, integration *s
 	}
 
 	integration.Id = uuid.NewV4().String()
-	return integration.Id, ds.storage.Upsert(ctx, integration)
+	error := ds.storage.Upsert(ctx, integration)
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore_impl.go:77]: err :=

File: central/imageintegration/datastore/datastore_impl.go [hunk 6]
```diff
@@ -76,16 +89,35 @@ func (ds *datastoreImpl) UpdateImageIntegration(ctx context.Context, integration
 		return sac.ErrResourceAccessDenied
 	}
 
-	return ds.storage.Upsert(ctx, integration)
+	error := ds.storage.Upsert(ctx, integration)
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore_impl.go:92]: err :=

File: central/imageintegration/datastore/datastore_impl.go [hunk 7]
```diff
@@ -76,16 +89,35 @@ func (ds *datastoreImpl) UpdateImageIntegration(ctx context.Context, integration
...
+	if error != nil {
+		return error
+	}
+	return ds.indexer.AddImageIntegration(integration)
 }
 
 // RemoveImageIntegration is pass-through to the underlying store.
 func (ds *datastoreImpl) RemoveImageIntegration(ctx context.Context, id string) error {
-	if ok, err := imageIntegrationSAC.WriteAllowed(ctx); err != nil {
+	if err := ds.storage.Delete(ctx, id); err != nil {
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore_impl.go:101]: why is sac removed?

File: central/imageintegration/datastore/datastore_impl.go [hunk 8]
```diff
@@ -76,16 +89,35 @@ func (ds *datastoreImpl) UpdateImageIntegration(ctx context.Context, integration
...
+func (ds *datastoreImpl) buildIndex(ctx context.Context) error {
+	if features.PostgresDatastore.Enabled() {
+		return nil
+	}
+	iis, err := ds.storage.GetAll(ctx)
+	log.Infof("[STARTUP] Found %d Image Integrations to be indexed", len(iis))
+	if err != nil {
+		return err
+	}
+	error := ds.indexer.AddImageIntegrations(iis)
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore_impl.go:116]: err :=

Comment (connorgorman) [reply central/imageintegration/datastore/datastore_impl.go:116]: flip the cases:
if err != nil { return err }
print log
return nil

File: central/imageintegration/search/searcher_impl.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+func (ds *searcherImpl) Search(ctx context.Context, q *v1.Query) ([]search.Result, error) {
+	return ds.searcher.Search(ctx, q)
+}
+
+func (ds *searcherImpl) Count(ctx context.Context, q *v1.Query) (int, error) {
+	return ds.searcher.Count(ctx, q)
+}
+
+// Format the search functionality of the indexer to be filtered (for sac) and paginated.
+func formatSearcher(unsafeSearcher blevesearch.UnsafeSearcher) search.Searcher {
```
Comment (connorgorman) [inline central/imageintegration/search/searcher_impl.go:42]: @md2119 I don't think we need this for a new image integration searcher right?

File: central/imageintegration/datastore/singleton.go [hunk 10]
```diff
@@ -22,44 +23,33 @@ var (
...
-	if !env.OfflineModeEnv.BooleanSetting() && len(integrations) == 0 {
+	if !env.OfflineModeEnv.BooleanSetting() && len(iis) == 0 {
 		// Add default integrations
 		for _, ii := range store.DefaultImageIntegrations {
 			utils.Must(storage.Upsert(ctx, ii))
 		}
 	}
 }
 
-func initializeManager(ctx context.Context) {
```
Comment (connorgorman) [inline central/imageintegration/datastore/singleton.go:36]: why is this removed?

File: central/enrichment/singleton.go [hunk 11]
```diff
@@ -53,6 +57,18 @@ func initialize() {
 	en = New(datastore.Singleton(), ie)
 	cf = fetcher.SingletonManager()
 	manager = newManager(imageintegration.Set(), ne, cf)
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowAllAccessScopeChecker())
```
Comment (connorgorman) [inline central/enrichment/singleton.go:60]: should make this a function like before

**Review by connorgorman — CHANGES_REQUESTED** [review]
See comments

**Comment by md2119** [comment]
@daynewlee What is this PR for?

File: central/cluster/datastore/datastore_impl.go [hunk 12]
```diff
@@ -152,6 +154,23 @@ func (ds *datastoreImpl) UpdateClusterStatus(ctx context.Context, id string, sta
 	return ds.clusterStorage.Upsert(ctx, cluster)
 }
 
+func (ds *datastoreImpl) removeClusterImageIntegrations(ctx context.Context, cluster *storage.Cluster) {
+
```
Comment (RTann) [inline central/cluster/datastore/datastore_impl.go:158]: nit: remove this newline

File: central/imageintegration/datastore/datastore.go [hunk 13]
```diff
@@ -16,17 +25,50 @@ var (
...
+		return nil
+	}
+	return ds
+}
+
+// NewForTestOnly returns an instance of DataStore only for tests.
+func NewForTestOnly(t *testing.T, storage store.Store, indexer index.Indexer, searcher search.Searcher) DataStore {
```
Comment (RTann) [inline central/imageintegration/datastore/datastore.go:58]: Instead of this, in the test, you can just do
```
ds := &datastoreImpl{
	storage:           storage,
	indexer:           indexer,
	formattedSearcher: searcher,
}
```

directly. Unless this is used in a different package as well?

File: central/imageintegration/datastore/datastore.go [hunk 14]
```diff
@@ -16,17 +25,50 @@ var (
...
+	ds := &datastoreImpl{
+		storage:           storage,
+		indexer:           indexer,
+		formattedSearcher: searcher,
+	}
+	return ds
+}
+
+// GetTestPostgresDataStore provides a datastore connected to postgres for testing purposes.
+func GetTestPostgresDataStore(t *testing.T, pool *pgxpool.Pool) (DataStore, error) {
```
Comment (RTann) [inline central/imageintegration/datastore/datastore.go:69]: It's usually not advised to put a test function in a non-test file. Where is this used?

**Review by RTann — COMMENTED** [review]
Can you rename the title of this PR to something like:
`ROX-9398: Delete image integrations upon cluster deletion`

Also, be sure to fill out the PR's description with the relevant information requested

**Review by rhybrillou — COMMENTED** [review]
Since you're changing the behaviour of the cluster removal function, I think it would be good to test that image integrations are deleted when there is any associated with the cluster.
The remaining remarks are nits and should not prevent the change from being merged.

**Review by rhybrillou — COMMENTED** [review]
Thanks for addressing my comments.
From my point of view, this looks OK to merge. Please make sure the comments of the other reviewers are addressed and they approve the change before merging.

File: central/imageintegration/datastore/singleton.go [hunk 15]
```diff
@@ -22,44 +23,33 @@ var (
...
+		indexer = index.New(globalindex.GetGlobalTmpIndex())
 	}
-	ad = New(storage)
-
-	ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowAllAccessScopeChecker())
-	initializeDefaultIntegrations(ctx, storage)
-	initializeManager(ctx)
+	initializeDefaultIntegrations(storage)
+	searcher := search.New(storage, indexer)
+	ad = New(storage, indexer, searcher)
```
Comment (RTann) [inline central/imageintegration/datastore/singleton.go:52]: an we change the name for this? I don't know what `ad` was meant to mean originally, so now might be a good time to give it a better name

File: central/imageintegration/search/searcher_impl.go [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+		}
+		// The result may not exist if the object was deleted after the search
+		if !exists {
+			continue
+		}
+		imageIntegrationList = append(imageIntegrationList, singleImageIntegration)
+	}
+
+	protoResults := make([]*v1.SearchResult, 0, len(imageIntegrationList))
+	for i, policy := range imageIntegrationList {
```
Comment (RTann) [inline central/imageintegration/search/searcher_impl.go:70]: nit: these aren't policies :)

File: central/imageintegration/search/searcher_impl.go (hunk 9 above)
Comment (RTann) [reply central/imageintegration/search/searcher_impl.go:43]: @md2119 bump ^

File: central/imageintegration/datastore/singleton.go (hunk 10 above)
Comment (RTann) [reply central/imageintegration/datastore/singleton.go:36]: Yeah keeping this here meant there would be an import cycle. We determined we can move this to `enrichment/singleton.go` and it should effectively be equivalent to keeping it here

File: central/cluster/datastore/datastore_impl_test.go [hunk 17]
```diff
@@ -127,8 +129,9 @@ func (suite *ClusterDataStoreTestSuite) SetupTest() {
 	suite.clusters.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.healthStatuses.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.indexer.EXPECT().AddClusters(nil).Return(nil)
-
+	suite.imageIntegrationDataStore = imageIntegrationDatastoreMocks.NewMockDataStore(suite.mockCtrl)
 	var err error
+
```
[the remark this answers is not in the document]
Comment (RTann) [reply central/cluster/datastore/datastore_impl_test.go:146]: Bump ^

File: central/enrichment/singleton.go (hunk 11 above)
Comment (RTann) [reply central/enrichment/singleton.go:59]: you can just initialize this from within `initializeManager` since it's only used there

File: central/imageintegration/datastore/datastore_impl_test.go [hunk 18]
```diff
@@ -320,13 +334,22 @@ func (suite *ImageIntegrationDataStoreTestSuite) TestEnforcesRemove() {
...
 	suite.NoError(err, "expected no error trying to write with permissions")
 
+	suite.indexer.EXPECT().DeleteImageIntegration(gomock.Any()).Return(nil)
 	integration = suite.storeIntegration("jdgbfdkjh2")
 
 	err = suite.datastore.RemoveImageIntegration(suite.hasWriteIntegrationsCtx, integration.GetId())
 	suite.NoError(err, "expected no error trying to write with permissions")
 }
+
+func (suite *ImageIntegrationDataStoreTestSuite) TestSearch() {
```
Comment (RTann) [inline central/imageintegration/datastore/datastore_impl_test.go:350]: This really just checks if `Search` is called. I think it'd be more beneficial to test the search process by creating a real searcher and indexer and then searching for a real element by ClusterID which you have added.

Comment (RTann) [reply central/imageintegration/datastore/datastore_impl_test.go:350]: Check out https://github.com/stackrox/stackrox/blob/master/pkg/bolthelper/bolt.go#L53 and https://github.com/stackrox/stackrox/blob/master/central/globalindex/bleveindex.go#L81

File: central/enrichment/singleton.go [hunk 19]
```diff
@@ -52,7 +56,24 @@ func initialize() {
 	ne = nodeEnricher.New(nodeCVESuppressor, metrics.CentralSubsystem)
 	en = New(datastore.Singleton(), ie)
 	cf = fetcher.SingletonManager()
+	initializeManager()
+}
+
+func initializeManager() {
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowAllAccessScopeChecker())
```
Comment (connorgorman) [inline central/enrichment/singleton.go:63]: use sac.WithAllAccess(context.Background())

File: central/imageintegration/datastore/datastore.go [hunk 20]
```diff
@@ -22,11 +32,50 @@ type DataStore interface {
...
-	return &datastoreImpl{
-		storage: storage,
+func New(imageIntegrationStorage store.Store, indexer index.Indexer, searcher search.Searcher) DataStore {
+	ds := &datastoreImpl{
+		storage:           imageIntegrationStorage,
+		indexer:           indexer,
+		formattedSearcher: searcher,
 	}
+
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(), sac.AllowAllAccessScopeChecker())
```
Comment (connorgorman) [inline central/imageintegration/datastore/datastore.go:48]: sac.WithAllAccess

File: central/imageintegration/datastore/datastore.go (hunk 4 above)
Comment (connorgorman) [reply central/imageintegration/datastore/datastore.go:51]: I think we can just use log.Fatalf here and combine the two lines

File: central/imageintegration/datastore/singleton.go (hunk 10 above)
Comment (connorgorman) [reply central/imageintegration/datastore/singleton.go:36]: 👍

File: central/imageintegration/search/searcher_impl.go [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/search/blevesearch"
+	"github.com/stackrox/rox/pkg/search/paginated"
+	"github.com/stackrox/rox/pkg/search/sortfields"
+)
+
+var (
+	defaultSortOption = &v1.QuerySortOption{
```
Comment (connorgorman) [inline central/imageintegration/search/searcher_impl.go:20]: No need for default sort option

File: central/imageintegration/search/searcher_impl.go [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+// Format the search functionality of the indexer to be filtered (for sac) and paginated.
+func formatSearcher(unsafeSearcher blevesearch.UnsafeSearcher) search.Searcher {
+	safeSearcher := blevesearch.WrapUnsafeSearcherAsSearcher(unsafeSearcher)
+	transformedSortFieldSearcher := sortfields.TransformSortFields(safeSearcher, imageIntegrationMapping.OptionsMap)
+	paginatedSearcher := paginated.Paginated(transformedSortFieldSearcher)
+	return paginated.WithDefaultSortOption(paginatedSearcher, defaultSortOption)
+}
+
+// Search retrieves SearchResults from the indexer and storage
+func (ds *searcherImpl) SearchImageIntegrations(ctx context.Context, q *v1.Query) ([]*v1.SearchResult, error) {
```
Comment (connorgorman) [inline central/imageintegration/search/searcher_impl.go:51]: Since this is not required for global search, do we need to return v1.SearchResult or can we just return the internal form which is simpler? Like this format: 
```Search(ctx context.Context, q *v1.Query) ([]search.Result, error)```

**Review by connorgorman — COMMENTED** [review]
Looking good, couple small things then give me a quick ping and I'll go back over again

**Review by connorgorman — APPROVED** [review]
